BubbleScan-AI/Custom.py

```diff
@@ -0,0 +1,130 @@
+        os.makedirs(aligned_folder, exist_ok=True)
+
+        for page_number, page in enumerate(pdf_document):
+            image_filename = f"CustomImage_{page_number + 1}.jpg"
+            image_path = os.path.join(aligned_folder, image_filename)
+
+            # Convert each page to an image and save
+            original_pix = page.get_pixmap(matrix=fitz.Identity, colorspace=fitz.csRGB)
+            scale_x = 1689 / original_pix.width
+            scale_y = 2186 / original_pix.height
+            matrix = fitz.Matrix(scale_x, scale_y)
+            pix = page.get_pixmap(matrix=matrix, colorspace=fitz.csRGB)
+            pix.save(image_path)
+            print(f"Extracted {image_filename}")
+
+        pdf_document.close()
+
+    def template_matching(self):
+        """
+        Align each page with the template for consistency in ROI extraction.
+        """
+        print("------Template Matching for Custom Sheets------")
+        template = cv2.imread(self.template_path)
+        aligned_folder = os.path.join(self.output_folder, "customAligned")
+        output_folder = os.path.join(self.output_folder, "customROIs")
+
+        image_files = [f for f in os.listdir(aligned_folder) if f.endswith(('.png', '.jpg', '.jpeg'))]
+
+        for image_file in sorted(image_files):
+            image_path = os.path.join(aligned_folder, image_file)
+            image = cv2.imread(image_path)
+
+            # Implement custom alignment here if necessary
+            aligned_image = self.align_image(image, template)
+
+            # Save aligned image if alignment is done
+            if aligned_image is not None:
+                aligned_path = os.path.join(output_folder, image_file)
+                cv2.imwrite(aligned_path, aligned_image)
+                print(f"Aligned and saved {image_file}")
+
+    def extractROIs(self):
+        """Extract regions of interest for each custom sheet image."""
+        print("------Extracting ROIs for Custom Sheet Images------")
+
+        aligned_images_folder = os.path.join(self.output_folder, "customAligned")
+        roi_folder_base = os.path.join(self.output_folder, "customROIs")
+        os.makedirs(roi_folder_base, exist_ok=True)
+
+        # List all images in the customAligned folder
+        image_files = [f for f in os.listdir(aligned_images_folder) if f.lower().endswith(('.png', '.jpg', '.jpeg'))]
+
+        for image_file in sorted(image_files):
+            image_path = os.path.join(aligned_images_folder, image_file)
+            print(f"Processing file: {image_file} for ROI extraction")
+
+            # Create a unique folder for each image's ROIs in customROIs
+            image_name = os.path.splitext(image_file)[0]
+            roi_folder = os.path.join(roi_folder_base, image_name)
+            os.makedirs(roi_folder, exist_ok=True)
+
+            # Extract and save ROIs based on updated layout
+            first_column, second_column, key_id, student_id = self.crop_roi(image_path, roi_folder)
```

Comment:
nice job splitting the first column and key id since they are not entirely aligned